Sources/Compiler/Types/TypeProtocol.swift

```diff
@@ -1,16 +1,69 @@
 import Utils
 
-/// A protocol describing the API of a type's associated value.
-public protocol TypeProtocol {
+/// A protocol describing the API of a Val type.
+public protocol TypeProtocol: Hashable {
 
   /// A set of flags describing recursive properties.
   var flags: TypeFlags { get }
 
+  /// Apply `transform(_:)` on the types that are part of `self`.
+  func transformParts(_ transformer: (AnyType) -> TypeTransformAction) -> Self
+
 }
 
 extension TypeProtocol {
 
   /// Returns whether the specified flags are raised on this type.
   public subscript(fs: TypeFlags) -> Bool { flags.contains(fs) }
 
+  /// Returns this type transformed with `transformer`.
+  ///
+  /// This method visits the structure of the type and calls `transformer` on each type composing
+  /// that structure. The result of the call substitutes the visited type. If `transformer` returns
```

Comment:
```suggestion
  /// that structure. The payload of the result is substituted for the visited type. If `transformer` returns
```

Except even that isn't quite accurate, because in the case of `stepInto` it's the recursively-transformed payload.  Happy to work with you on this.